### PR TITLE
MOTS-578: Roles and Permission screen

### PR DESCRIPTION
**Pull request by pmuchowski**

### CODE REVIEW

File: src/main/java/org/motechproject/mots/web/UserController.java [hunk 1]
```diff
@@ -203,4 +203,72 @@ public UserProfileDto getUserProfile(Principal principal) {
...
+  @ResponseStatus(HttpStatus.OK)
+  @ResponseBody
+  public RoleDto getRole(@PathVariable("id") UUID id) {
+
+    UserRole role = userService.getRole(id);
+    return roleMapper.toDto(role);
+  }
+
+  /**
+   * Finds roles matching all of the provided parameters.
```
Comment (jkondrat) [inline src/main/java/org/motechproject/mots/web/UserController.java]: missing javadoc

File: src/webapp/js/components/sidebar.js [hunk 2]
```diff
@@ -241,20 +241,36 @@ class SideBar extends Component {
       <ul className="nav nav-second-level">
         { hasAuthority(MANAGE_USERS_AUTHORITY, MANAGE_INCHARGE_USERS_AUTHORITY) &&
         <li className="border-none">
-          <Link to="/users/new" onClick={this.props.hideMenuSmart}>
+          <Link to="/users/users/new" onClick={this.props.hideMenuSmart}>
```
Comment (jkondrat) [inline src/webapp/js/components/sidebar.js]: why do we need to change the url to 'users/users'?

File: src/webapp/js/reducers/tables-data/tables_data_initial_state.js [hunk 3]
```diff
@@ -8,4 +8,6 @@ export default {
   locationListPages: 1,
   userListPages: 1,
   roles: [],
+  rolesList: [],
```
Comment (jkondrat) [inline src/webapp/js/reducers/tables-data/tables_data_initial_state.js]: why do we have a list of roles and a "rolesList"?

File: src/main/java/org/motechproject/mots/web/UserController.java (hunk 1 above)
Comment (pmuchowski) [reply src/main/java/org/motechproject/mots/web/UserController.java]: @jkondrat this is the javadoc, so what do you mean?

File: src/main/resources/db/changelog/20181019_1441.mysql.sql [hunk 4]
```diff
@@ -0,0 +1,89 @@
+--liquibase formatted sql
+
+--changeset pmuchowski:1539952921216-1
+CREATE TABLE user_permission (id VARCHAR(255) NOT NULL, created_date datetime NULL, updated_date datetime NULL, display_name VARCHAR(255) NULL, name VARCHAR(255) NOT NULL, readonly BIT DEFAULT 0 NOT NULL);
```
[the remark this answers is not in the document]
Comment (pmuchowski) [reply src/main/resources/db/changelog/20181019_1441.mysql.sql]: the class is name UserPermission and the pattern we are using to name the tables is to use snake case class name

File: src/main/resources/db/changelog/20181019_1441.mysql.sql [hunk 5]
```diff
@@ -0,0 +1,89 @@
...
+
+--changeset pmuchowski:1539952921216-2
+ALTER TABLE user_role_permissions ADD permission_id VARCHAR(255) NOT NULL;
+
+--changeset pmuchowski:1539952921216-3
+ALTER TABLE user_role ADD readonly BIT DEFAULT 0 NOT NULL;
```
Comment (pmuchowski) [inline src/main/resources/db/changelog/20181019_1441.mysql.sql]: see above

File: src/webapp/js/reducers/tables-data/tables_data_initial_state.js (hunk 3 above)
Comment (pmuchowski) [reply src/webapp/js/reducers/tables-data/tables_data_initial_state.js]: roles is the list of all roles used in selects and rolesList is the result of the search (in the roles table)

File: src/main/java/org/motechproject/mots/constants/DefaultPermissions.java [hunk 6]
```diff
@@ -0,0 +1,49 @@
+package org.motechproject.mots.constants;
+
+@SuppressWarnings("PMD.AvoidDuplicateLiterals")
+public class DefaultPermissions {
```
[the remark this answers is not in the document]
Comment (pmuchowski) [reply src/main/java/org/motechproject/mots/constants/DefaultPermissions.java]: the "ROLE_" prefix is required by spring security

File: src/webapp/js/components/sidebar.js (hunk 2 above)
Comment (jkondrat) [reply src/webapp/js/components/sidebar.js]: so 'users/roles' is the list of user's roles? could we have the list of users under '/users' (the parent)?

File: src/webapp/js/reducers/tables-data/tables_data_initial_state.js (hunk 3 above)
Comment (jkondrat) [reply src/webapp/js/reducers/tables-data/tables_data_initial_state.js]: let's change name of 'rolesList' to 'filteredRoles'

File: src/webapp/js/components/sidebar.js (hunk 2 above)
Comment (pmuchowski) [reply src/webapp/js/components/sidebar.js]: no, the users/roles is the list of all roles, also there are users/user/new to create new user and users/users/{userId} to edit user and for roles users/role/new and users/role/{roleId}

**Pull request merged**